theurgy.lic

```diff
@@ -2,185 +2,453 @@
+      }
+    end
+    if DRStats.circle > 2
+      communes << {
+        name: :tamsine,
+        method: :commune_tamsine
+      }
+    end
+    if DRStats.circle > 60
+      communes << {
+        name: :truffenyi,
+        method: :commune_truffenyi,
+        items: [
+          {
+            name: 'Taffelberries',
+            shop: @data['taffelberry_shop'],
+            purchase_method: :buy_taffelberries,
+          },
+        ],
+      }
+    end
+    if DRStats.circle > 8
+      communes << {
+        name: :kertigen,
+        method: :commune_kertigen,
+        items: [
+          {
+            name: 'oil',
+            shop: @data['oil_shop'],    
+          },
+        ],
+      }
+    end
   end
 
-  def get_from_container(item)
-    bput "get #{item} from my #{@theurgy_supply_container}", 'You get', 'I could not find', 'What were you referring to'
+  def annotate_item_presence(data)
+    data.each { |datum| datum[:items] ||= [] }
+    items = data.map { |d| d[:items] }.flatten.uniq { |i| i[:name] }
+    item_presence = {}
+    items.each { |i| item_presence[i[:name]] = on_hand?(i) }
+
+    data.each do |datum|
+      datum[:items].each { |i| i[:on_hand] = item_presence[i[:name]] }
+    end
+  end
+  
+  def rituals_remain?
+    !@rituals.empty?
   end
 
-  ######################## COMMUNES ########################
+  def on_hand?(item)
+    if item[:name] == 'oil'
+      inside_container?('holy oil') || inside_container?('some oil')
+    elsif item[:any_container]
+      exists?(item[:name])
+    else
+      inside_container?(item[:name])
+    end
+  end
 
-  def comm_eluned
-    fput "stow #{GameObj.right_hand.noun}" if GameObj.right_hand.name != 'Empty'
-    fput "stow #{GameObj.left_hand.noun}" if GameObj.left_hand.name != 'Empty'
+  def inside_container?(item_name)
+    inside?(item_name, @theurgy_supply_container)
+  end
 
-    forage? 'dirt'
+  def put_in_container(item_name)
+    bput("put #{item_name} in my #{@theurgy_supply_container}",
+         'You put', 'What were you referring to')
+  end
 
-    get_from_container(@water_holder)
-    bput('commune eluned', 'completed this commune too recently', 'You grind some dirt in your fist')
-    put_in_container(@water_holder)
-    fput('drop dirt') if [left_hand, right_hand].include?('dirt')
+  def get_from_container(item_name)
+    bput("get #{item_name} from my #{@theurgy_supply_container}",
+         'You get', 'I could not find', 'What were you referring to')
+  end
+    
+  def have_holy_water?
+    inside?('holy water', @water_holder)    
   end
 
-  def comm_tamsine
-    if inside?('holy water', @water_holder)
-      pause 1
-      waitrt?
-      get_from_container(@water_holder)
-      fput "sprinkle #{@water_holder} on #{checkname}"
-      bput('commune tamsine', 'completed this commune too recently', 'You feel warmth spread throughout your body')
-      put_in_container(@water_holder)
-    else
-      Flags.delete('theurgy-tamsine')
-      nil
+  def buy_supplies
```

Comment:
This could be a user setting as well to help us non-f2p chars make better use of our time. This could go in as a todo.

Reply:
Agreed that this is a good idea.  I'll add this as a todo and it will go in one of my next PRs.  This complicates things a bit as it requires computing the total amount we expect to spend ahead of time (which we really ought to do anyway).